sunscreen_runtime/src/builder.rs

```diff
@@ -0,0 +1,648 @@
+    pub fn public_inputs<I>(mut self, inputs: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        ZkpProgramInput: From<T>,
+    {
+        self.public_inputs
+            .extend(inputs.into_iter().map(ZkpProgramInput::from));
+        self
+    }
+
+    /// Verify that `self.proof` satisfies `self.program`; see
+    /// [`runtime.verify()`][GenericRuntime::verify].
+    ///
+    /// # Remarks
+    /// Will error if the underlying `verify` call errors, or if a proof has not yet been
+    /// supplied to the builder. That is, you must call [`Self::proof`] before calling this
+    /// function.
+    pub fn verify(self) -> Result<()> {
+        let proof = self.proof.ok_or_else(|| {
+            BuilderError::user_error(
+                "You must supply a proof to the verification builder before calling `verify`",
+            )
+        })?;
+        self.runtime.verify(
+            self.program,
+            proof,
+            self.public_inputs,
+            self.constant_inputs,
+        )
+    }
+}
+
+#[cfg(feature = "linkedproofs")]
+pub use linked::*;
+
+#[cfg(feature = "linkedproofs")]
+mod linked {
+    use super::BuilderError;
+
+    use std::{borrow::Cow, sync::Arc};
+
+    use logproof::{
+        bfv_statement::{self, BfvMessage, BfvProofStatement, BfvWitness, StatementParams},
+        rings::{SealQ128_1024, SealQ128_2048, SealQ128_4096, SealQ128_8192},
+        Bounds, LogProofProverKnowledge,
+    };
+    use seal_fhe as seal;
+    use seal_fhe::SecurityLevel;
+    use sunscreen_compiler_common::{Type, TypeName};
+    use sunscreen_math::ring::{BarrettBackend, BarrettConfig, Zq};
+    use sunscreen_zkp_backend::{
+        bulletproofs::{BulletproofsBackend, BulletproofsFieldSpec},
+        FieldSpec,
+    };
+
+    use crate::{
+        marker, BFVEncryptionComponents, Ciphertext, CompiledZkpProgram, GenericRuntime,
+        LinkedProof, NumCiphertexts, Params, Plaintext, PublicKey, Result, Sdlp,
+        SealSdlpProverKnowledge, TryIntoPlaintext, ZkpProgramInput,
+    };
+
+    /// A trait for sharing a plaintext with a ZKP program.
+    pub trait ShareWithZkp: NumCiphertexts {
+        /// The associated ZKP input type when this plaintext is shared with a ZKP program.
+        type ZkpType<F: FieldSpec>: TypeName;
+
+        /// The number of nonzero coefficients to share between the SDLP and ZKP.
+        ///
+        /// Note that when plaintexts polynomials are shared with ZKP programs, the ZKP program
+        /// assumes the coefficients form a 2s complement encoding. Without this bound, all
+        /// coefficients (up to the lattice degree `N`) are shared and thus the ZKP circuit has to
+        /// compute `2^N`. Of course, each ZKP field element is limited by its underlying
+        /// representation (e.g. a 256-bit scalar for Bulletproofs). Thus, any encoding should keep
+        /// this limit this in mind.
+        ///
+        /// This number should be less than 256.
```

Comment:
Why less than 256? What if they want to share an entire BFV 1024 plaintext?

Reply:
Well, I suppose for this trait that isn't a _strict_ requirement in general. But currently all the integer encodings use coefficients multiplied by powers of 2. So attempting the decoding at `N = 256` would result in `2^N` overflowing. I can reword this to make it more clear.

Reply:
Oh I was thinking a 256 degree polynomial. Whoops.